src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenDeconstructTests.cs

```diff
@@ -4773,21 +4933,229 @@ static void Main()
-            //Diagnostic(ErrorCode.ERR_NameNotInContext, "_").WithArguments("_").WithLocation(6, 10),
-            // (6,13): error CS0103: The name '_' does not exist in the current context
-            //         (_, _) = (new C(), new C());
-            //Diagnostic(ErrorCode.ERR_NameNotInContext, "_").WithArguments("_").WithLocation(6, 13)
-            //CompileAndVerify(comp, expectedOutput: "CC");
+            comp.VerifyDiagnostics();
+            CompileAndVerify(comp, expectedOutput: "CC");
         }
 
+        [Fact]
+        public void DeclareAndUseLocalInDeconstruction()
+        {
+            var source =
+@"
+class C
+{
+    static void Main()
+    {
+        (var x, x) = (1, 2);
+        (y, var y) = (1, 2);
+    }
+}
+";
+
+            var comp = CreateCompilationWithMscorlib(source, options: TestOptions.DebugExe, references: s_valueTupleRefs);
+            // mixing declaration and expressions isn't supported yet
+            comp.VerifyDiagnostics(
+                // (6,17): error CS1031: Type expected
+                //         (var x, x) = (1, 2);
+                Diagnostic(ErrorCode.ERR_TypeExpected, "x").WithLocation(6, 17),
+                // (7,10): error CS1031: Type expected
+                //         (y, var y) = (1, 2);
+                Diagnostic(ErrorCode.ERR_TypeExpected, "y").WithLocation(7, 10),
+                // (6,17): error CS0128: A local variable or function named 'x' is already defined in this scope
+                //         (var x, x) = (1, 2);
+                Diagnostic(ErrorCode.ERR_LocalDuplicate, "x").WithArguments("x").WithLocation(6, 17),
+                // (7,17): error CS0128: A local variable or function named 'y' is already defined in this scope
+                //         (y, var y) = (1, 2);
+                Diagnostic(ErrorCode.ERR_LocalDuplicate, "y").WithArguments("y").WithLocation(7, 17)
+                );
+        }
+
+        [Fact]
+        public void OutVarAndUsageInDeconstructionAssignment()
+        {
+            var source =
+@"
+class C
+{
+    static void Main()
+    {
+        (M(out var x).P, x) = (1, x);
+        System.Console.Write(x);
+    }
+    static C M(out int i) { i = 42; return new C(); }
+    int P { set { System.Console.Write($""Written {value}. ""); } }
+}
+";
+
+            var comp = CreateCompilationWithMscorlib(source, options: TestOptions.DebugExe, references: s_valueTupleRefs);
+            comp.VerifyDiagnostics();
+            CompileAndVerify(comp, expectedOutput: "Written 1. 42");
+        }
+
+        [Fact]
+        public void OutDiscardInDeconstructionAssignment()
+        {
+            var source =
+@"
+class C
+{
+    static void Main()
+    {
+        (M(out var _).P, _) = (1, 2);
+        System.Console.Write(x);
+    }
+    static C M(out int i) { i = 42; return new C(); }
+    int P { set { System.Console.Write($""Written {value}. ""); } }
+}
+";
+            // PROTOTYPE(wildcards) Crash
```

Comment:
Yes (casting exception), the binding of an out var needs to be updated to handle a `DiscardedVariableSyntax`.